src/storage/src/compaction/rate_limit.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use crate::error::{CompactionRateLimitedSnafu, Result};
+
+pub trait RateLimitToken {
+    /// Releases the token.
+    /// ### Note
+    /// Implementation should guarantee the idempotency.
+    fn try_release(&self);
```

Comment:
Looks like the user should call `try_release` by hand, what about using `Drop` to release the token implicitly?

Reply:
When token gets released,
https://github.com/GreptimeTeam/greptimedb/blob/3beb07193e32cd87e78077ad6b6dc6452e1717a9/src/storage/src/compaction/scheduler.rs#L224

rather than using a scoped variable to control the release before notify `task_notify`, I'd prefer explicitly call `try_release`. 

Another consideration is, those token may involve some async rpc invocation in the future (for example, if we need to control simultanenously running tasks num across whole cluster based on a counter in metasrv).